# Employee Exit Surveys Analysis
In this project, we will work with exit surveys from employees of the [Department of Education, Training and Employment](https://en.wikipedia.org/wiki/Department_of_Education_and_Training_(Queensland)) (DETE) and the Technical and Further Education (TAFE) Institute in Queensland, Australia. The data can be found [here](https://data.gov.au/dataset/ds-qld-fe96ff30-d157-4a81-851d-215f2a0fe26d/details?q=exit%20survey) (the original TAFE exit data survey is no longer available).

The **GOALS** of our project are to answer these questions:
- Are employees who only worked for the institutes for a short period of time resigning due to some kind of dissatisfaction? What about the employees who have been there longer?
- Are younger employees resigning due to some kind of dissatisfation? What about older employees?

A data dictionary wasn't provided with the dataset. For this project, we will use our general knowledge to define the columns.

Below is a preview of a couple columns we'll work with from the `dete_survey.csv`:
* `ID`: Id used to identify the participant of the survey
* `SeparationType`: The reason why the person's employment ended
* `Cease Date`: The year or month the person's ended
* `DETE Start Date`: The year the person began employment with the DETE

Below is a preview of a couple columns we'll work with from the `tafe_survey.csv`:
* `record_ID`: Id used to identify the participant of the survey
* `Reason for ceasing employment`: The reason why the person's employment ended
* `LengthofServiceOverall. Overall Length of Service at Institute (in years)`: The length of the person's employment (in years)

## 1. Exploring and Cleaning the Data
Let's start exploring our datasets.

In [2]:
import pandas as pd
import numpy as np

dete_survey = pd.read_csv('dete_survey.csv')
tafe_survey = pd.read_csv('tafe_survey.csv')

### 1.1. DETE Survey Data

In [3]:
dete_survey.info()
dete_survey.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 822 entries, 0 to 821
Data columns (total 56 columns):
 #   Column                               Non-Null Count  Dtype 
---  ------                               --------------  ----- 
 0   ID                                   822 non-null    int64 
 1   SeparationType                       822 non-null    object
 2   Cease Date                           822 non-null    object
 3   DETE Start Date                      822 non-null    object
 4   Role Start Date                      822 non-null    object
 5   Position                             817 non-null    object
 6   Classification                       455 non-null    object
 7   Region                               822 non-null    object
 8   Business Unit                        126 non-null    object
 9   Employment Status                    817 non-null    object
 10  Career move to public sector         822 non-null    bool  
 11  Career move to private sector        822 non-

ID                    SeparationType Cease Date DETE Start Date  \
0   1             Ill Health Retirement    08/2012            1984   
1   2  Voluntary Early Retirement (VER)    08/2012      Not Stated   
2   3  Voluntary Early Retirement (VER)    05/2012            2011   
3   4         Resignation-Other reasons    05/2012            2005   
4   5                    Age Retirement    05/2012            1970   

  Role Start Date                                      Position  \
0            2004                                Public Servant   
1      Not Stated                                Public Servant   
2            2011                               Schools Officer   
3            2006                                       Teacher   
4            1989  Head of Curriculum/Head of Special Education   

  Classification              Region                      Business Unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     Employment Status  ...  Kept informed  Wellness programs  \
0  Permanent Full-time  ...              N                  N   
1  Permanent Full-time  ...              N                  N   
2  Permanent Full-time  ...              N                  N   
3  Permanent Full-time  ...              A                  N   
4  Permanent Full-time  ...              N                  A   

   Health & Safety  Gender          Age  Aboriginal  Torres Strait  South Sea  \
0                N    Male        56-60         NaN            NaN        NaN   
1                N    Male        56-60         NaN            NaN        NaN   
2                N    Male  61 or older         NaN            NaN        NaN   
3                A  Female        36-40         NaN            NaN        NaN   
4                M  Female  61 or older         NaN            NaN        NaN   

   Disability  NESB  
0         NaN   Yes  
1         NaN   NaN  
2         NaN   NaN  
3         NaN   NaN  
4         NaN   NaN  

[5 rows x 56 columns]

In the `dete_survey` dataset, we have 822 rows and 56 columns. However we won't use all of the columns in this analysis, and focused only to the first four columns: `ID`, `SeparationType`, `Cease Date`, and `DETE Start Date`. The `Cease Date` and `DETE Start Date` are object they while they should be datetime type, we will convert them later.

In [28]:
dete_survey['Cease Date'].value_counts()

2012          344
2013          200
01/2014        43
12/2013        40
Not Stated     34
09/2013        34
06/2013        27
07/2013        22
10/2013        20
11/2013        16
08/2013        12
05/2013         7
05/2012         6
04/2014         2
07/2014         2
08/2012         2
04/2013         2
02/2014         2
11/2012         1
09/2010         1
2010            1
2014            1
07/2012         1
09/2014         1
07/2006         1
Name: Cease Date, dtype: int64

On `Cease Date` column, we found that there is no missing (NaN) value in the column. However, there are 34 entries of `Not Stated` value. We will treat them as missing values, and we have options what to do with them:
- Drop the rows with `Not Stated` values in `Cease Date` column.
- Replace with other value that approximately represent the data.
- Leave them as is.

First, we can use `pd.read_csv()` function to specify values that should be represented as `NaN`. In this process, we will also drop other columns that aren't useful to our analysis.

In [31]:
dete_survey = pd.read_csv('dete_survey.csv', na_values='Not Stated')
dete_survey_updated = dete_survey.drop(dete_survey.columns[28:49],
                                      axis=1)
dete_survey_updated.head()

ID                    SeparationType Cease Date  DETE Start Date  \
0   1             Ill Health Retirement    08/2012           1984.0   
1   2  Voluntary Early Retirement (VER)    08/2012              NaN   
2   3  Voluntary Early Retirement (VER)    05/2012           2011.0   
3   4         Resignation-Other reasons    05/2012           2005.0   
4   5                    Age Retirement    05/2012           1970.0   

   Role Start Date                                      Position  \
0           2004.0                                Public Servant   
1              NaN                                Public Servant   
2           2011.0                               Schools Officer   
3           2006.0                                       Teacher   
4           1989.0  Head of Curriculum/Head of Special Education   

  Classification              Region                      Business Unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-AO7      Central Office  Corporate Strategy and Peformance   
2            NaN      Central Office               Education Queensland   
3        Primary  Central Queensland                                NaN   
4            NaN          South East                                NaN   

     Employment Status  ...  Work life balance  Workload  None of the above  \
0  Permanent Full-time  ...              False     False               True   
1  Permanent Full-time  ...              False     False              False   
2  Permanent Full-time  ...              False     False               True   
3  Permanent Full-time  ...              False     False              False   
4  Permanent Full-time  ...               True     False              False   

   Gender          Age  Aboriginal  Torres Strait  South Sea  Disability  NESB  
0    Male        56-60         NaN            NaN        NaN         NaN   Yes  
1    Male        56-60         NaN            NaN        NaN         NaN   NaN  
2    Male  61 or older         NaN            NaN        NaN         NaN   NaN  
3  Female        36-40         NaN            NaN        NaN         NaN   NaN  
4  Female  61 or older         NaN            NaN        NaN         NaN   NaN  

[5 rows x 35 columns]

Now we have 35 columns instead of 56, and the `Not Stated` values have been replaced with `NaN` values.

### 1.2. TAFE Survey Data
Next, let's look at the TAFE dataset.

In [32]:
tafe_survey.info()
tafe_survey.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 72 columns):
 #   Column                                                                                                                                                         Non-Null Count  Dtype  
---  ------                                                                                                                                                         --------------  -----  
 0   Record ID                                                                                                                                                      702 non-null    float64
 1   Institute                                                                                                                                                      702 non-null    object 
 2   WorkArea                                                                                                                                  

Record ID                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   WorkArea  CESSATION YEAR Reason for ceasing employment  \
0  Non-Delivery (corporate)          2010.0              Contract Expired   
1  Non-Delivery (corporate)          2010.0                    Retirement   
2       Delivery (teaching)          2010.0                    Retirement   
3  Non-Delivery (corporate)          2010.0                   Resignation   
4       Delivery (teaching)          2010.0                   Resignation   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

   ...  \
0  ...   
1  ...   
2  ...   
3  ...   
4  ...   

  Workplace. Topic:Does your workplace promote a work culture free from all forms of unlawful discrimination?  \
0                                                Yes                                                            
1                                                Yes                                                            
2                                                Yes                                                            
3                                                Yes                                                            
4                                                Yes                                                            

  Workplace. Topic:Does your workplace promote and practice the principles of employment equity?  \
0                                                Yes                                               
1                                                Yes                                               
2                                                Yes                                               
3                                                Yes                                               
4                                                Yes                                               

  Workplace. Topic:Does your workplace value the diversity of its employees?  \
0                                                Yes                           
1                                                Yes                           
2                                                Yes                           
3                                              

In [35]:
tafe_survey['CESSATION YEAR'].value_counts(dropna=False)

2011.0    268
2012.0    235
2010.0    103
2013.0     85
NaN         7
2009.0      4
Name: CESSATION YEAR, dtype: int64

On TAFE dataset, we have 702 entries with 72 columns. And we know that missing values are already properly defined as `NaN`. However we noticed that most of the column names are long and will make our code writing difficul, we will simplify them.

Prior to that, we will drop the irrelevant columns first:

In [37]:
tafe_survey_updated = tafe_survey.drop(tafe_survey.columns[17:66], axis=1)
tafe_survey_updated.head()

Record ID                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   WorkArea  CESSATION YEAR Reason for ceasing employment  \
0  Non-Delivery (corporate)          2010.0              Contract Expired   
1  Non-Delivery (corporate)          2010.0                    Retirement   
2       Delivery (teaching)          2010.0                    Retirement   
3  Non-Delivery (corporate)          2010.0                   Resignation   
4       Delivery (teaching)          2010.0                   Resignation   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

   ... Contributing Factors. Study Contributing Factors. Travel  \
0  ...                         NaN                          NaN   
1  ...                           -                       Travel   
2  ...                           -                            -   
3  ...                           -                       Travel   
4  ...                           -                            -   

  Contributing Factors. Other Contributing Factors. NONE  \
0                         NaN                        NaN   
1                           -                          -   
2                           -                       NONE   
3                           -                          -   
4                           -                          -   

  Gender. What is your Gender? CurrentAge. Current Age  \
0                       Female                  26  30   
1                          NaN                     NaN   
2                          NaN                     NaN   
3                          NaN                     NaN   
4                         Male                  41  45   

  Employment Type. Employment Type Classification. Classification  \
0              Temporary Full-time            Administration (AO)   
1                              NaN                            NaN   
2                              NaN                            NaN   
3                              NaN                            NaN   
4              Permanent Full-time        Teacher (including LVT)   

  LengthofServiceOverall. Overall Length of Service at Institute (in years)  \
0                                                1-2                          
1                     

### 1.4. Standardizing the Column Names
Next, let's turn our attention to the column names. Each dataframe contains many of the same columns, but the column names are different. Below are some of the columns we'd like to use for our final analysis.

|dete_survey|tafe_survey|Definition|
|:---------:|:---------:|:--------:|
|ID|Record ID|An id used to identify the prticipant of the survey|
|SeparationType|Reason for ceasing employment|The reason why the participant's employmeny ended|
|Cease Date|CESSATION YEAR|The year or month the participant's employment ended|
|DETE Start Date||The year the participant began employment with the DETE|
||LengthofServiceOverall.Overall Length of Service at Institute (in years)|the length of the person's employment (in years)|
|Age|CurrentAge.Current Age|The age of the participant|
|Gender|Gender.What is your Gender?|The gender of the participant|

Because later we will combine the two datasets, we'll have to standardize the column names. We can use the `DataFrame.columns` attribute along with vertorized string methods to update all of the columns at once.

In [55]:
# Renaming the DETE dataset
dete_survey_updated.columns = (dete_survey_updated.columns.
                              str.lower().str.strip().
                              str.replace(' ', '_'))

# Renaming the TAFE dataset
renaming = {'Record ID':'id', 'CESSATION YEAR':'cease_date',
            'Reason for ceasing employment':'separationtype',
            'Gender. What is your Gender?':'gender',
            'CurrentAge. Current Age':'age',
            'Employment Type. Employment Type':'employment_status',
            'Classification. Classification':'position',
            'LengthofServiceOverall. Overall Length of Service at Institute (in years)':'institute_service',
            'LengthofServiceCurrent. Length of Service at current workplace (in years)':'role_service'}
tafe_survey_updated.rename(renaming, axis=1, inplace=True)

In [56]:
# Checking the renamed DETE dataset
dete_survey_updated.head(2)

id                    separationtype cease_date  dete_start_date  \
0   1             Ill Health Retirement    08/2012           1984.0   
1   2  Voluntary Early Retirement (VER)    08/2012              NaN   

   role_start_date        position classification          region  \
0           2004.0  Public Servant        A01-A04  Central Office   
1              NaN  Public Servant        AO5-AO7  Central Office   

                       business_unit    employment_status  ...  \
0  Corporate Strategy and Peformance  Permanent Full-time  ...   
1  Corporate Strategy and Peformance  Permanent Full-time  ...   

   work_life_balance  workload  none_of_the_above  gender    age  aboriginal  \
0              False     False               True    Male  56-60         NaN   
1              False     False              False    Male  56-60         NaN   

   torres_strait  south_sea  disability  nesb  
0            NaN        NaN         NaN   Yes  
1            NaN        NaN         NaN   NaN  

[2 rows x 35 columns]

In [57]:
# Checking the renamed TAFE dataset
tafe_survey_updated.head(2)

id                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   

                   WorkArea  cease_date    separationtype  \
0  Non-Delivery (corporate)      2010.0  Contract Expired   
1  Non-Delivery (corporate)      2010.0        Retirement   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   

   ... Contributing Factors. Study Contributing Factors. Travel  \
0  ...                         NaN                          NaN   
1  ...                           -                       Travel   

  Contributing Factors. Other Contributing Factors. NONE  gender     age  \
0                         NaN                        NaN  Female  26  30   
1                           -                          -     NaN     NaN   

     employment_status             position institute_service role_service  
0  Temporary Full-time  Administration (AO)               1-2          1-2  
1                  NaN                  NaN               NaN          NaN  

[2 rows x 23 columns]

### 1.5. Isolating the Data
We have renamed the columns that we'll use in our analysis. Next, let's remove more of the data we don't need.

In [63]:
print(dete_survey_updated['separationtype'].value_counts())
print('\n')
print(tafe_survey_updated['separationtype'].value_counts())

Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: separationtype, dtype: int64


Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: separationtype, dtype: int64


If we look at the unique values in the `separationtype` columns in each dataframe, we'll see that each contains a couple of different separation types. For this project, we'll only analyze survey respondents who *resigned*, so their separation type contains the sring `Resignation`.

Note that `dete_survey_updated` dataframe contains multiple separation types with the string `Resignation`:
- Resignation-Other reasons
- Resignation-Other employer
- Resignation-Move overseas/interstate

We will have to account for each of these variations so we don't unintentionally drop data!

In [66]:
# Isolate the TAFE dataframe which contains "Resignation" in separationtype column
tafe_resignations = tafe_survey_updated[tafe_survey_updated['separationtype'] == 'Resignation'].copy()

In [72]:
# Isolate the TAFE dataframe which contains "Resignation" in separationtype column
pat = r"Resignation"
dete_resignations = dete_survey_updated[dete_survey_updated['separationtype'].str.contains(pat)].copy()

For the TAFE dataframe we only isolate the datapoints which values in the `separationtype` column **equal to** `Resignation`. In contrast, for the DETE dataframe we have to use regular expression (regex) to capture every value which **contains** string `Resignation`, regardless the characters followed after.

### 1.6. Checking for Wrong Data
In this step, we'll focus on verifying that the years in the `cease_date` and `date_start_date` columns to make sense.
- Since the `cease_date` is the last of the person's employment and the `dete_start_date` is the person's first year of employment, it wouldn't make sense to have years after the current date.
- Given that most people in this field start working in their 20s, it's also unlikely that the `dete_start_date` was before the year 1940.

If we have many years higher than the current date or lower than 1940, we wouldn't want to continue with our analysis, because it could mean there's something very wrong with the data. if there are a small amount of values that are unrealistically high ir low, we can remove them.